### ISSUE TITLE
Package Microsoft.Extensions.Logging.Abstractions requires netstandard 2.0

### ISSUE THREAD

**Issue by JamesRandall** [aspnet/Logging#687]
Would this package not be more useful if it targeted a lower version of netstandard (it's still compatible with 1.3) thus allowing it to be used in a broader range of scenarios. Per Microsoft's [own guidance](https://docs.microsoft.com/en-us/dotnet/standard/net-standard):

_In general, we recommend you to target the lowest version of .NET Standard possible._ 

Unless my understanding is wrong by targeting netstandard 2.0 doesn't this force any other package that relies upon it to also make the leap to 2.0? When MS started splitting out these foundational components into abstractions and implementation packages I thought it was great to see some reference interfaces being defined in this space while at the same time allowing for theoretically many implementations - we've arguably lost something if they are going to needlessly push on to the latest broadest netstandard implementation.

Admittedly there is currently only a 2.0 implementation of the abstractions and so reality it's going to be running inside a runtime that supports netstandard 2.0 but surely this "platform raising" is better left as far up towards runtime selection as possible to maximise flexibility lower down the dependency graph.

As an example if I had a set of packages that I wanted to support, say, UWP it would be fairly trivial to implement a simple logger on top of the abstractions. But by forcing all packages that reference this to netstandard 2.0 I've got a much tougher job.

**Comment by davidfowl** [comment]
@JamesRandall What can't you target the 1.x abstractions?

**Comment by JamesRandall** [comment]
I can work round it by directly referencing 2.0.0 of the abstractions in the 1 of 5 packages I have that requires this (it has a dependency on Entity Framework Core which has a >= 2.0.0 dependency on Microsoft.Extensions.Logging).

In that world I can leave the other 4 packages targeting netstandard 1.3 and logging abstractions 1.x and have the package that depends on Entity Framework Core target netstandard 2.0 and logging abstractions 2.x. 

However then, within a related set of packages, I have a mix of major version changes of dependencies that ostensibly follow a semver model  - by definition that's dangerous (a major version bump being used to indicate the making of incompatible API changes) and my packages, or their consumers, are eventually going to run fowl of this.

Whereas if the logging abstractions follow the recommended best practice (that makes an awful lot of sense to me) and target the lowest compatible netstandard version then, I would argue, everything is playing nicely with both the semver model and the netstandard approach.

netstandard, and the best practice advice around it, coupled to semver make an awful lot of sense to me in a world of multiple runtimes and a more granular approach to framework delivery - but aren't we on a bit of a hiding to nothing if the core packages released by Microsoft don't play by these guidelines?

And so a reciprocal question, I'm genuinely curious, is there a reason .Net Core and related packages aren't targeting the minimum netstandard version that they can?

**Comment by davidfowl** [comment]
> netstandard, and the best practice advice around it, coupled to semver make an awful lot of sense to me in a world of multiple runtimes and a more granular approach to framework delivery - but aren't we on a bit of a hiding to nothing if the core packages released by Microsoft don't play by these guidelines?

The best practice still makes sense in general. You generally want to target the lowest netstandard you can. We made an executive decision to reset the baseline for our packages given we were shipping a major version (and removing TFMs is a breaking change). NET Standard 2.0 has lots of benefits like cleaning up the package dependency graph (faster install times etc). It also simplifies the compatibility matrix we had to test since we did the reset. 

I don't expect this will be a common occurrence. .NET Standard 2.0 was a big change and we wanted to align with it.

**Comment by JamesRandall** [comment]
I guess the concern is that improvements come along all the time but runtimes move at different cadences. If major version bumps result in runtimes being left behind  by package management following this approach then I'm not sure what we've achieved with .NET Standard.

As you suggest hopefully this won't happen often and therefore won't turn out to be the case. 

I've bumped my packages along to .NET Standard 2.0 as it's the cleanest solution available to me given package purpose (primarily distributed services, Azure), audience (small), and the discussion above.

Thanks for the info and back and forth.

**Comment by peteraritchie** [comment]
I agree with this issue.  
Bumping the version *of this package* to 2.0 makes sense in the context that going to .NET Standard 2.0, because *that* is a breaking change.  **But**, this seems to conflict with the referenced guidance but also seems to introduce potential versioning issues as well as violating the principle of least surprise.  It's certainly adds needless friction to developers in the wild.

Versioning examples: 

- What if a change needed in the .NET Standard 1.1 supported version of the library (Abstractions 1.2)? That would make a new version (say 1.3) have a version smaller than and *older* version.
- What if a change like a above (a change required for the latest ,NET Standard 1.1 supported version) is in fact a breaking change?  Does that them become 3.0 because the "breaking change" in 2.0 would in fact be an aggregate breaking change? 😕 

Example friction:

I'm creating a class library targeting .NET Standard 1.2 because I need to support existing code that doesn't yet support .NET Standard/Core 2.0. That library would like to support a logging abstraction, so I attempt to add the package to my class library.  The NuGet package manager in Visual Studio proceeds to try to install Abstractions 2.0 and, of course, barfs.  As a dev I *now* need to know that *maybe* a prior version would in fact support the .NET Standard that I need, then go looking for past versions that do that 🤞 .  *This assumes that .NET Standard versions prior to 2.0 will be supported for a least a while in the future.*

This strategy doesn't make long-term sense for any level of support for .NET Standards <2.0.

**Comment by vpenades** [comment]
I also agree... I have packages and code that can't be bumped to 2.0... so I'm stuck now with M.E.L.Abstractions 1.1.2.

I can foresee several consequences of this breaking change in the long run:

- Packages that do require to stay at the lowest possible net.standard version, to remove M.E.L.Abstractions altoghether.
- Third party logging providers, to be forced to provide updates for both M.E.L.Abstracions 2.0, and for 1.1.2 ( for those that want to stick with a lower version )

In both cases, bumping the version defeats the original purpose of M.E.L.Abstracions to be a foundation for an universal logging system.

**Comment by JamesRandall** [comment]
As the general issue remains and it is causing others similar concerns and difficulties it doesn't seem appropriate for this to be closed by me and so am reopening.

**Comment by vpenades** [comment]
So, could it be possible to bump Abstractions back to net.standard 1.1 ? or it is already out of reach?
